torch/distributed/pipelining/stage.py

```diff
@@ -1258,23 +1258,13 @@ def __init__(
         dw_builder: Optional[Callable[[], Callable[..., None]]] = None,
     ):
         super().__init__(submodule, stage_index, num_stages, device, group, dw_builder)
-        self.submod.to(self.device)
-        # When we materialize the model partition on cuda, we call reset_parameters() if it is available
-        self.inputs: List[torch.Tensor] = []
-        self.outputs: List[torch.Tensor] = []
-
-        self.inputs = _create_empty_tensors(input_args, device)
-
-        if output_args is None:
-            logger.info("output_args not provided, performing forward using input_args")
-            self.outputs = self.submod(*self.inputs)
-            # create buffers for the output so that the data is in the correct
-            # shape in order to use in p2p op (send)
-            self.outputs = _create_empty_tensors(self.outputs, device)
-        else:
-            self.outputs = _create_empty_tensors(output_args, device)
-
-        self._configure_outputs_meta(tuple(self.outputs))
+        self.inputs: List[torch.Tensor] = None
+        self.inputs_meta = (
+            (input_args,) if isinstance(input_args, torch.Tensor) else input_args
+        )
+        self._configure_outputs_meta(
```

Comment:
This will make `output_args` required (right now it is optional), otherwise initialization will fail right? To have a better transition, can we still run `self.submod()` with the input args and leave the output_args to be false. This will fail if the model is not on the same device as the input, but that is okay and we can say we expect them to be both on meta device.

Reply:
you're right that this makes output_args required.  I like your suggestion.  If you're willing to let it be an error if users pass real inputs, I would propose to assert the module and inputs are on the same device, and then compute 'outputs_meta' based on inputs.

I was weighing whether to make it a requirement to pass input/module on meta device, but I suppose that is too restrictive.  I do want to ensure that outputs_meta is stored on meta to avoid wasting memory.  So maybe if the user passes cuda model/inputs I will convert the output to meta.  Wdyt?

Reply:
ok, i added output-shape inference back in, but i made it refuse to do inference on non-meta device, so some user code might still have to switch their inputs to meta.  is this ok, or should i make it work for cuda?

Reply:
It probably might be easier to just let it work for cuda if the users already have it on cuda. Then you wouldn't have to fix all the tests as you mentioned above. I think the main fix of the PR can just be removing the `.to(device)` in init which you did, and with that removal now it is up to the users to make sure that `model(input_args)` works if they pass it in.

Reply:
Hmm, can I say that I prefer the previous version better? i.e. removal.
(See also my comments about on relaxing device constraint, which requires removal of dryrun.)

Reply:
Hmm, I'm in the other boat, I think keeping the dry run makes more sense (as a temporary measure) until we get the lazy init working, otherwise we have to update all the tests to pass in both input_args and output_args and also requires users to do so as well.

Reply:
Yeah, that's hard. Maybe at least add `with torch.no_grad()`? 
I am actually not sure what side effect running a module has. Saving grad context, marking requires_grad?

Reply:
Also, in `forward_maybe_with_...` we have special logic for DDP/FSDP modules, would this dryrun's escape be an issue? (Or are we just lucky?)

Reply:
i'm going to leave this for later. its not any worse with this PR and we plan to remove it when adding lazy shape inference.  sound ok?

Reply:
Yep, sounds good.